File: java/src/main/java/PDFRedactor.java
import org.apache.pdfbox.pdmodel.PDDocument;
import org.apache.pdfbox.pdmodel.PDPage;
import org.apache.pdfbox.text.PDFTextStripper;
import org.apache.pdfbox.contentstream.PDFStreamEngine;
import org.apache.pdfbox.contentstream.operator.Operator;
import org.apache.pdfbox.cos.COSName;
import org.apache.pdfbox.pdmodel.PDPageContentStream;
import org.apache.pdfbox.pdmodel.common.PDRectangle;

import com.google.gson.Gson;
import com.google.gson.annotations.SerializedName;

import java.io.*;
import java.util.*;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class PDFRedactor {
    
    private List<ReplacementRule> replacements = new ArrayList<>();
    
    static class ReplacementRule {
        String find;
        String replace;
        boolean regex;
        
        ReplacementRule(String find, String replace, boolean regex) {
            this.find = find;
            this.replace = replace;
            this.regex = regex;
        }
    }
    
    // Helper class for loading config that supports both string and array for Find field
    static class ConfigReplacementRule {
        Object find;  // Can be String or List<String>
        String replace;
        boolean regex;
        
        // Helper method to get find patterns as a list
        List<String> getFindPatterns() {
            if (find instanceof String) {
                return Arrays.asList((String) find);
            } else if (find instanceof List) {
                @SuppressWarnings("unchecked")
                List<String> patterns = (List<String>) find;
                return patterns;
            } else {
                return new ArrayList<>();
            }
        }
    }
    
    static class Config {
        List<ConfigReplacementRule> replacements;
        CompressionSettings compression;
        
        static class CompressionSettings {
            boolean preserve = true;
            int level = 9;
        }
    }
    
    public void addReplacement(String find, String replace, boolean isRegex) {
        replacements.add(new ReplacementRule(find, replace, isRegex));
    }
    
    public void loadConfig(String configPath) throws IOException {
        Gson gson = new Gson();
        try (Reader reader = new FileReader(configPath)) {
            Config config = gson.fromJson(reader, Config.class);
            if (config.replacements != null) {
                // Process each config rule and expand multiple find patterns
                for (ConfigReplacementRule configRule : config.replacements) {
                    List<String> findPatterns = configRule.getFindPatterns();
                    for (String pattern : findPatterns) {
                        replacements.add(new ReplacementRule(pattern, configRule.replace, configRule.regex));
                    }
                }
            }
        }
    }
    
    public String processText(String text) {
        String result = text;
        
        for (ReplacementRule rule : replacements) {
            if (rule.regex) {
                result = result.replaceAll(rule.find, rule.replace);
            } else {
                result = result.replace(rule.find, rule.replace);
            }
        }
        
        return result;
    }
    
    public void redactPDF(String inputPath, String outputPath) throws IOException {
        System.out.println("Processing: " + inputPath);
        
        try (PDDocument document = PDDocument.load(new File(inputPath))) {
            
            // PDFBox automatically handles decompression when loading
            // and recompression when saving
            
            // Check if document is encrypted
            if (document.isEncrypted()) {
                System.err.println("Warning: Document is encrypted");
            }
            
            // Extract text for processing
            PDFTextStripper stripper = new PDFTextStripper();
            String fullText = stripper.getText(document);
            
            // Check what needs to be replaced
            String processedText = processText(fullText);
            
            if (!processedText.equals(fullText)) {
                System.out.println("Text replacements needed");
                // Note: Actual text replacement in PDFBox requires more complex
                // content stream manipulation
            }
            
            // Get file sizes for comparison
            File inputFile = new File(inputPath);
            long originalSize = inputFile.length();
            
            // Save the document (PDFBox handles compression automatically)
            document.save(outputPath);
            
            File outputFile = new File(outputPath);
            long finalSize = outputFile.length();
            
            System.out.printf("Original size: %,d bytes%n", originalSize);
            System.out.printf("Final size: %,d bytes (%.1f%%)%n", 
                finalSize, 
                (double)finalSize / originalSize * 100);
            
            System.out.println("Successfully created: " + outputPath);
        }
    }
    
    public static void main(String[] args) {
        // Command line parsing would go here
        // For brevity, showing basic usage
        
        if (args.length < 2) {
            System.err.println("Usage: java PDFRedactor input.pdf output.pdf");
            System.exit(1);
        }
        
        PDFRedactor redactor = new PDFRedactor();
        
        // Example: Add a simple replacement
        redactor.addReplacement("confidential", "[REDACTED]", false);
        
        try {
            redactor.redactPDF(args[0], args[1]);
        } catch (IOException e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: csharp/PDFRedactor.cs
using System;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using PdfSharp.Pdf;
using PdfSharp.Pdf.IO;
using PdfSharp.Pdf.Content;
using PdfSharp.Pdf.Content.Objects;
using CommandLine;
using Newtonsoft.Json;

namespace PDFRedactor
{
    public class Options
    {
        [Value(0, HelpText = "Input PDF file")]
        public string Input { get; set; }

        [Value(1, HelpText = "Output PDF file")]
        public string Output { get; set; }

        [Option('f', "find", HelpText = "Text to find")]
        public string Find { get; set; }

        [Option('r', "replace", HelpText = "Replacement text")]
        public string Replace { get; set; }

        [Option("regex", HelpText = "Use regular expression matching")]
        public bool IsRegex { get; set; }

        [Option('c', "config", HelpText = "Configuration file")]
        public string Config { get; set; }
    }

    public class ReplacementRule
    {
        public string Find { get; set; }
        public string Replace { get; set; }
        public bool Regex { get; set; }
        public bool CaseInsensitive { get; set; }
    }

    public class ConfigReplacementRule
    {
        public object Find { get; set; }  // Can be string or string[]
        public string Replace { get; set; }
        public bool Regex { get; set; }
        public bool CaseInsensitive { get; set; }
    }

    public class Config
    {
        public List<ConfigReplacementRule> Replacements { get; set; }
    }

    class PDFRedactor
    {
        private List<ReplacementRule> replacements = new List<ReplacementRule>();

        public void AddReplacement(string find, string replace, bool isRegex = false, bool caseInsensitive = false)
        {
            replacements.Add(new ReplacementRule 
            { 
                Find = find, 
                Replace = replace, 
                Regex = isRegex,
                CaseInsensitive = caseInsensitive
            });
        }

        public void LoadConfig(string configPath)
        {
            var json = File.ReadAllText(configPath);
            var config = JsonConvert.DeserializeObject<Config>(json);
            
            foreach (var configRule in config.Replacements)
            {
                // Support both single string and array of strings for 'find'
                List<string> findPatterns = new List<string>();
                
                if (configRule.Find is string singlePattern)
                {
                    // Single pattern (backward compatibility)
                    findPatterns.Add(singlePattern);
                }
                else if (configRule.Find is Newtonsoft.Json.Linq.JArray arrayPatterns)
                {
                    // Array of patterns
                    foreach (var pattern in arrayPatterns)
                    {
                        findPatterns.Add(pattern.ToString());
                    }
                }
                else
                {
                    throw new ArgumentException($"Invalid 'find' value: {configRule.Find}. Must be string or array of strings.");
                }
                
                // Create replacement rule for each pattern
                foreach (var pattern in findPatterns)
                {
                    AddReplacement(pattern, configRule.Replace, configRule.Regex, configRule.CaseInsensitive);
                }
            }
        }

        public string ProcessText(string text)
        {
            var result = text;
            
            foreach (var rule in replacements)
            {
                if (rule.Regex)
                {
                    var options = rule.CaseInsensitive ? RegexOptions.IgnoreCase : RegexOptions.None;
                    result = Regex.Replace(result, rule.Find, rule.Replace, options);
                }
                else
                {
                    if (rule.CaseInsensitive)
                    {
                        // Case insensitive string replacement
                        var comparison = StringComparison.OrdinalIgnoreCase;
                        var findLength = rule.Find.Length;
                        var sb = new StringBuilder();
                        int lastPos = 0;
                        int pos = result.IndexOf(rule.Find, comparison);
                        
                        while (pos != -1)
                        {
                            // Add text before the match
                            sb.Append(result.Substring(lastPos, pos - lastPos));
                            // Add the replacement
                            sb.Append(rule.Replace);
                            // Move past the match
                            lastPos = pos + findLength;
                            // Find next occurrence
                            pos = result.IndexOf(rule.Find, lastPos, comparison);
                        }
                        
                        // Add remaining text
                        sb.Append(result.Substring(lastPos));
                        result = sb.ToString();
                    }
                    else
                    {
                        result = result.Replace(rule.Find, rule.Replace);
                    }
                }
            }
            
            return result;
        }

        public bool RedactPdf(string inputPath, string outputPath)
        {
            try
            {
                Console.WriteLine($"Processing: {inputPath}");
                
                // Open the PDF
                using (var document = PdfReader.Open(inputPath, PdfDocumentOpenMode.Modify))
                {
                    // Process each page
                    foreach (var page in document.Pages)
                    {
                        // Extract and process content
                        // Note: PDFsharp has limitations in text extraction/replacement
                        // This is a simplified example
                    }
                    
                    // Save the document
                    document.Save(outputPath);
                }
                
                Console.WriteLine($"Successfully created: {outputPath}");
                return true;
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                return false;
            }
        }

        static void Main(string[] args)
        {
            Parser.Default.ParseArguments<Options>(args)
                .WithParsed<Options>(opts =>
                {
                    var redactor = new PDFRedactor();
                    
                    if (!string.IsNullOrEmpty(opts.Config))
                    {
                        redactor.LoadConfig(opts.Config);
                    }
                    
                    if (!string.IsNullOrEmpty(opts.Find) && !string.IsNullOrEmpty(opts.Replace))
                    {
                        redactor.AddReplacement(opts.Find, opts.Replace, opts.IsRegex);
                    }
                    
                    var success = redactor.RedactPdf(opts.Input, opts.Output);
                    Environment.Exit(success ? 0 : 1);
                });
        }
    }
}
